### DISCUSSION TITLE
Connection drops with with Modbus Proxy / Huawei Solar / EVCC

### DISCUSSION THREAD

**Discussion by Kugelfang666** [wlcrs/huawei_solar#699] [Connecting to the inverter]
Hi all,

I have quite a complex setup:

Inverter 1 (with Battery + S-Dongle)
Inverter 2 (cascaded with Inverter 1)
Huawei Wallbox

Home Assistant with ModbusProxy connected to S-Dongle on Inverter 1

Huawei-Solar-Integration connected to Modbus Proxy
EVCC connected to Modbus Proxy

I see frequent connection drops in the form of spikes in the sensor values in all curves. These are going away the moment I disable EVCC. Could it be that there is some blockage or so by both integrations querying the S-Dongle? 

Anyone that has some recommendations on how to optimize my setup? I’m not so sure about the right settings for connection time and connection timeout on the Modbus proxy

![IMG_0497](https://github.com/wlcrs/huawei_solar/assets/3289654/c064c521-9ed1-487c-ace5-74bbc3c89798)
![IMG_0498](https://github.com/wlcrs/huawei_solar/assets/3289654/0d567ed6-791e-45a5-bb93-55071a9e31d1)

**Comment by tmenguy** [comment]
Sorry no clue :) ... but this is the setup I was envisioning soon to move to EVCC (or do all by automation inside HA) ... so interested in your problem.

**Comment by wlcrs** [comment]
I'm surprised you are getting 0 values. I would expect some connection drops to occur, but HA and the integration should handle it gracefully like "value missing" (and thus ignored) instead of logging a 0.

Feel free to create a bug report where you include debug logs during which these 0 values manifest themselves + the exact timestamps of these 0 values in your history (cfr. https://www.home-assistant.io/integrations/history/#exporting-data-from-the-history-panel )

**Comment by Kugelfang666** [comment]
I think the 0-drops occurs since i have a calculated value in the graph and this drops to 0 when the source sensor from the inverter goes to unavailable. Sorry if this caused some confusion

**Reply by wlcrs** [reply]
Make sure to use the `availability` feature in your computed sensor to prevent this from happening.

**Comment by reyntjensw** [comment]
@Kugelfang666  did you find a way to get this working?

**Reply by Kugelfang666** [reply]
Yes, but it was a Journey

Ultimately I removed the cascade. 

Now I have:

Inverter 1 with sDongle 1, Smartmeter, Luna200

Inverter 2 with sDongle 2

Both are connected separately to the same FusionSolar plant

In EVCC I created Two PV installations so both sDongles are being queried via ModbusTCP

My HomeAssitant server has two USB-WiFi adapters. Each is connected to one inverter via the near field WiFi. This way this integration is bypassing the sDongle and directly querying the inverter. 

This was more reliable for me and mitigates the issue with the single client per dongle.

**Reply by CommanderROR9** [reply]
Interesting...
I have just one Inverter and am querying it directly (Wifi Dongle Thingy) but as soon as I use the Modbus Proxy, all hell breaks loose. I get lots of 0 Values (same Reason, calculated Sensor)  and loads of dropped connections, broken pipe, timeout etc. from the Modbus Proxy, even with just the Home Assistant Integration talking to it. If I add evcc as well, then everything breaks down immediately.

**Comment by DanielBarie** [comment]
Same issue here. Direct connection to dongle (10-KTL) works. But I need the dongle data to be accessible to another client (uni-meter). Tried via HA Modbus proxy. Uni-meter works via proxy. Everything fails as soon as the Huawei_solar addon gets connected to the proxy (even uni-meter will lose connectivity). Tried the evcc proxy, pretty much the same result. So I think it is about this integration (maybe polling is too aggressive?).

**Reply by CommanderROR9** [reply]
Interestingly...this still holds true, even with a direct connection via the Inverter Hotspot. So even taking the Dongle out of the equation.
It doesn't break immediately, it can take hours, days or even weeks, but eventually it all crashes and burns with the Proxy.

**Comment by Kugelfang666** [comment]
Meanwhile a lot has changed: There is a rich discussion ongoing about different HW revisions of the dongle. I upgraded both my dongles to the faster HW version version and ever since they are MUCH more responsible and can handle more intense loads. Detailed discussion can be found here: 

https://www.photovoltaikforum.com/thread/247827-huawei-dongle-verbindungsprobleme/?pageNo=1

and 

https://github.com/evcc-io/evcc/discussions/22534

**Comment by DanielBarie** [comment]
Yes, but according to my experience that won't change the behavior when connected to a/the Modbus proxy.

Am 29. März 2026 21:45:52 UTC schrieb Kugelfang666 ***@***.***>:

**Reply by StefanW86** [reply]
I also have the problem. It works over weeks and then the connection is lost. Only restart of modbusproxy and evcc/ Home Assistant helps.

**Comment by mcamposv** [comment]
I had the same instability issues running a software Modbus proxy between HA and an EMMA. The root 
problem is that Modbus only allows one active client — when HA and EVCC both try to query at the 
same time, things break.

I ended up building a transparent Modbus TCP proxy on an ESP32 that handles the multiplexing in 
hardware, with its own reconnection logic independent of the cloud. Running stable in my setup. 
Might be worth a look if you're still hitting drops: https://github.com/mcamposv/modbusproxy

**Reply by CommanderROR9** [reply]
Interesting. I will have to check it out, though I admit that I don't really understand what you did. A Proxy is essentially just Software. Running it on an esp32 vs. on my powerful Proxmox Server as an Home Assistant Addon...I don't really "get" what differentiates it. But I will read your Github and see if I can follow.

**Reply by mcamposv** [reply]
The key difference isn't raw computing power — you're right that a Proxmox server is far more powerful than an ESP32. The value is independence and simplicity.

  A software proxy running inside Home Assistant (or as an addon on the same machine) shares the same failure domain: if HA restarts, updates, or crashes, your Modbus connection drops too. The ESP32 sits between
   your inverter and everything else, always on, with no dependency on HA, Proxmox, or any VM being healthy.

  The practical advantages for your use case specifically:

  Resilience: the proxy never restarts unless you tell it to. HA and EVCC can restart independently without interrupting the inverter connection.

  Quick swap: you can keep a pre-flashed backup ESP32 ready. If hardware fails, you're back online in under 10 minutes — no reinstalling addons or reconfiguring anything.

  Built-in Modbus sniffer: the device exposes a web dashboard where you can watch every Modbus transaction in real time. Given that your drops seem to happen when HA and EVCC query simultaneously, this could be
  very useful — you'd see exactly which request is causing the conflict and whether it's a timing issue or a protocol one.

  Regarding your connection drops specifically: the root cause you described (HA + EVCC competing for the single Modbus slot) is exactly the problem this proxy solves by serializing all requests through a single
   persistent connection to the inverter. Worth trying before adding more dongles.

**Comment by samumar82** [comment]
@mcamposv it has been 3 days that I've been trying to implent your peoject.
As I have no more LAN ports availble on the switch I would like to go with wifi and test latency etc.
I have an ESP32 S3 N16R8 spare but can0t figure it out what to change in the code, tried to vibe coding but nothing works.
I also tried with an ESP32 WROOM 32D but same results. 
Can you support? I don't have EMMA but a Sdongle connected via lan.
Thanks in advance!

**Comment by mcamposv** [comment]
Hello Rovin Ronin,

I think software should work fine on both boards (asuming you compile 
and deploy it correctly)

I can confirm that I had it running on ESP32 WROOM 32D for testing and 
it did correctly.

I would need more details on what is failing there. to try to help you.

without any data I can only thing on a couple of things.

First thing please try to go through the readme files I think I put the 
install procedure. but anyway I'll try to mumarize here.

you need to edit the right data on the file src/secrets.h before 
compiling/deploying.

When you fash the device for the first time. it willcreate a wifi 
hotspot where you can connect with your movile or laptop and it will 
show a config website.

where you can select your wifi and your details.  (then only think there 
is if you are using dhcp after connection you need to know the device ip 
addr. (I think it prints information on serial port)

if you put all values in the installation website it should work. 
(asuming that you have tcp modbus enabled on your inverter and no 
restrictions on ip address and there is no other client using that 
tcpmodbus  as inverter only allows one connection which should be our proxy.

Please send me more details and I'll try to help.

**Comment by samumar82** [comment]
Hello and thanks for your reply. 
I managed to make it work on the wroom 32D, for the S3 it's almost impossible without deeply change the code. 
After configuring the proxy o wasn't able to connect to the inverter and only later I read it better and found out the sdongle is not supported, only Emma and inverter directly.
In my case I need the dongle with lan connection for the scharger walbox to work properly. 
If you confirm it's not possible with the dongle I think I will have to implement it on proxmox. 
I was much happier with a separate solution who does not update overnight or reboot. 
The issue is that the proxy connects ok to the dongle but the the dongle does not get back any info from the inverter. :-(

**Comment by mcamposv** [comment]
hello,

sorry to hear that code needs a lot of change for the S3 I do not have 
any board or knowledge I just quickly checked with gemini and said it 
should work anyway you do not need that big board.


About the dongle compatibility I think that basically if you have the 
dongle you do not need.  the modbus proxy. the dongle already support 
multiple connections.

I have only tested it with the emma setup. which is my install (and I do 
not know why when you have the Emma you have to remove the dongle as 
emma takes control. (that was the root of my issues)

I have not tested talking directly to de inverter as again emma takes 
over.  but if someone has only the inverter is a mmater of using the 
scan tools to figure out what is the ID of the inverter in case it changes.

May be you can try that as well. you can try to run scans to the dongle 
ip and see if the inverter has a diferent ID

in the repo there are some scanning script in python to run directly for 
your laptop.

**Comment by samumar82** [comment]
I'm almost ok with a proxy in proxmox, will try to understand how to port
it to the esp32 once it works without any issues.

Unfortunately the dongle does support only one modbus connection even
though it is set to unrestricted.

This is an intended behaviour from Huawei, just one connection is
permitted. That's why people are struggling with proxies.

Il Gio 10 Set 2026, 20:06 Guybrush ***@***.***> ha scritto:

**Comment by samumar82** [comment]
Thanks for sharing the idea, I made it working with some mods I share with you.
In this repo https://github.com/samumar82/huawei-sunbridge-modbus you can find both the proxmox ct and the esp32 proxy.
Esp32 can run via lan if you have the possibility or via wifi.
Wifi set as fallback when lan is not detected/working.
